TableWriter 8: Refactor table writer APIs

gggrace14 opened the pull request:
Refactor HiveDataSink APIs and their usage in TableWriter.

Remove WriteProtocol and WriteProtocol registering, and
move the logic of WriteProtocol to connector-specific. We do not
support variant behaviors of the same connector at the moment.
Different connectors could have different logic of write & commit.

Add class HiveConfig and include
insert_existing_partitions_behavior

Need to follow 1) fileSize field population in PartitionUpdate json;
2) Constant vector of CommitContext is not flattened during
serialization.

mbasmanova commented on velox/connectors/WriteProtocol.h:
```diff
@@ -30,6 +30,11 @@ class WriterParameters {
   virtual ~WriterParameters() = default;
 };
 
+class ConnectorWriteInfo {
```
Why do we need a separate interface? Can the necessary information be included in ConnectorInsertTableHandle?

gggrace14 replied:
I'm planning to use ConnectorWriteInfo to pass the partition subdirectory (or partition name) to WriteProtocol::getWriterParameters(),  which is Hive connector specific.

We get the partition subdirectories on the fly when reading the inputs and parsing the individual records. Here is where I populate ConnectorWriteInfo in another PR, https://github.com/facebookincubator/velox/blob/feec0f6736c48297b9853d92a394aa9a656e9050/velox/connectors/hive/HiveDataSink.cpp#L212. ConnectorInsertTableHandle are passed from the coordinator and converted from the similar protocol structure, https://github.com/prestodb/presto/blob/48a7d5f477d7d1c462a0a95e5a174c63d9bb623b/presto-native-execution/presto_cpp/presto_protocol/presto_protocol.cpp#L6231-L6250.

mbasmanova replied:
Thank you for explaining. Would you document this interface?

mbasmanova commented on velox/connectors/hive/HiveDataSink.cpp:
```diff
@@ -70,17 +81,8 @@ std::unique_ptr<velox::dwrf::Writer> HiveDataSink::createWriter() {
-
-  auto writePath = fs::path(hiveWriterParameters->writeDirectory()) /
-      hiveWriterParameters->writeFileName();
+  auto writePath = fs::path(writerParameters->writeDirectory()) /
```
This change looks like a code movement without any logic change. Is this so? Why is it needed?

gggrace14 replied:
Yeah, it's just a code movement. I'm moving the `auto hiveWriterParameters = writeProtocol_->getWriterParameters()` out of this createWriter() class and to append(), and pass hiveWriterParameters to createWriter(WriterParameters). It's because in the next PR we need to pass partitionDirectory when calling writeProtocol_->getWriterParameters(). I feel it's more clean to pass WriterParameters to createWriter() compared w/ passing partitionDirectory.

mbasmanova commented on velox/connectors/hive/HiveWriteProtocol.h:
```diff
@@ -88,31 +96,50 @@ class HiveWriterParameters : public WriterParameters {
+      std::optional<std::string> partitionDirectory = std::nullopt)
+      : partitionDirectory_(std::move(partitionDirectory)) {}
+
+  ~HiveConnectorWriteInfo() override = default;
```
not needed

mbasmanova commented on velox/connectors/hive/HiveWriteProtocol.cpp:
```diff
@@ -53,29 +57,46 @@ HiveNoCommitWriteProtocol::getWriterParameters(
       connectorQueryCtx->taskId(),
       connectorQueryCtx->driverId(),
       makeUuid());
+  auto targetDirectory = hiveWriteInfo->partitionDirectory().has_value()
```
this code repeats twice; consider extracting a helper method

```
auto targetDirectory = makePartitionDirectory(hiveWriteInfo->partitionDirectory(), hiveTableWriteHandle->locationHandle()->targetPath());
```

gggrace14 replied:
Got it. Removing the repetition.

mbasmanova commented on velox/connectors/hive/HiveWriteProtocol.h:
```diff
@@ -31,6 +31,8 @@ class HiveWriterParameters : public WriterParameters {
 
   /// @param updateMode Write the files to a new directory, or append to an
   /// existing directory or overwrite an existing directory.
+  /// @param partitionName The name of partition if any that the writer writes
```
Would you expand the documentation to explain what partitionName is and how it is used in combination with targetDirectory and writeDirectory?

My understanding is partitionName is actually a path to partition subdirectory that looks like a=<a>/b=<b>/c=<c>, where a, b and c are partition keys. It is appended to targetDirectory and writeDirectory paths to create the actual path to write to.

gggrace14 replied:
Yes, exactly. partitionName is actually the partition subdirectory in string, in the format of a=/b=/c=. Revising the documentation.

mbasmanova reviewed (commented):
@gggrace14 Overall looks good. The PR description could be expanded to provide more explanation for the changes. Some comments below.

gggrace14 commented on velox/connectors/hive/HiveConfig.h:
```diff
@@ -0,0 +1,96 @@
+    return baseConfig->get<uint32_t>(kMaxPartitionsPerWriters, 100);
+  }
+
+  static InsertExistingPartitionsBehavior insertExistingPartitionsBehavior(
```
To validate and set the value, replicating logic https://github.com/prestodb/presto/blob/ff412ef4b745eb31549cc0abe25e885a63875f58/presto-hive/src/main/java/com/facebook/presto/hive/HiveClientConfig.java#L643-L650.

mbasmanova commented on velox/connectors/hive/HiveConfig.h:
```diff
@@ -0,0 +1,96 @@
+
+  /// Can new data be inserted into existing partitions or existing
+  /// unpartitioned tables.
+  static constexpr const char* FOLLY_NONNULL kImmutablePartitions =
```
We no longer need FOLLY_NONNULL annotations after #3468

gggrace14 commented:
Hi @mbasmanova Masha, I have some questions about session properties
1. How do HiveClientConfigs are passed to the Velox worker? I can see session properties go to connectorQueryCtx->config(). How about HiveClientConfigs? Session property uses "_" while HiveConfig uses "-". Ex., hive.immutable-partitions which doesn't have corresponding session property.
2. How are hive.xxx and prism_batch.xxx sessions of different catalogs are passed to Velox worker? They probably share the same connector key, hive?

mbasmanova commented on velox/docs/configs.rst:
```diff
@@ -2,8 +2,11 @@
 Configuration properties
 ========================
 
+Query Configuration Properties
```
I don't think we need this extra section. Do we?

mbasmanova replied:
"Query Configuration Properties" section is redundant. Let's remove.

mbasmanova commented on velox/docs/configs.rst:
```diff
@@ -164,4 +167,35 @@ data size exceeds this limit, otherwise it spills the partition with most data.
+files.
+
+
+Hive Configuration Properties
```
Hive Connector (drop Configuration Properties as this is redundant).

mbasmanova replied:
I still think this section should be named "Hive Connector".

gggrace14 replied:
Will do

mbasmanova commented on velox/docs/configs.rst:
```diff
@@ -164,4 +167,35 @@ data size exceeds this limit, otherwise it spills the partition with most data.
+Hive Configuration Properties
+-----------------------------
+
+Table Writer
```
I don't think we need this sub-section. Do we?

mbasmanova replied:
I still think we don't need this section.

gggrace14 replied:
Works for me

mbasmanova commented on velox/docs/configs.rst:
```diff
@@ -164,4 +167,35 @@ data size exceeds this limit, otherwise it spills the partition with most data.
+    * **Type:** ``boolean``
+    * **Default value:** ``true``
+
+Can new data be inserted into existing partitions or existing unpartitioned tables.
```
The property name "immutable_partitions" suggests that it applies only to partitioned tables, but the description suggests it applies to unpartitioned tables as well. Is this correct?

gggrace14 replied:
Yes, this immutable_partitions config also have effect on the path for unpartitioned table. Cross-reading the presto java oss code I got the impression that the unpartitioned table might be treated as single partitioned table.

mbasmanova commented on velox/docs/configs.rst:
```diff
@@ -164,4 +167,35 @@ data size exceeds this limit, otherwise it spills the partition with most data.
+    * **Type:** ``string``
+    * **Default value:** ``ERROR``
+
+Behavior on insert existing partitions.
```
Would you document all supported values and their semantics?

gggrace14 replied:
Done, added the docs

mbasmanova reviewed (commented):
@gggrace14 Ge, thank you for adding documentation for Hive Connector configuration properties. Some comments.

mbasmanova commented on velox/connectors/hive/HiveConfig.h:
```diff
@@ -0,0 +1,55 @@
+  static constexpr const char* kInsertExistingPartitionsBehavior =
+      "insert_existing_partitions_behavior";
+
+  static bool isImmutablePartitions(const Config* baseConfig);
```
Why baseConfig and not just config?

gggrace14 replied:
Corrected.

mbasmanova commented on velox/connectors/hive/HiveConfig.cpp:
```diff
@@ -0,0 +1,75 @@
+  auto strBehavior =
+      baseConfig->get<std::string>(kInsertExistingPartitionsBehavior);
+  if (strBehavior.has_value()) {
+    InsertExistingPartitionsBehavior behavior =
```
auto

gggrace14 commented on velox/connectors/hive/HiveConfig.h:
```diff
@@ -0,0 +1,55 @@
+    kOverwrite,
+  };
+
+  /// Can new data be inserted into existing partitions or existing
```
This is the most important session property for writing I think. So I suggest we carry the semantics. And it is shared by the worker and the coordinator. Presto coordinator has very complex logic around it, https://github.com/prestodb/presto/blob/82441f5ef3d6f38dc61428db20039d9dc95dae1f/presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java#L2097.

My understanding of the semantics is

> When set to ``APPEND``, Velox will append additional files to existing partitions. When set to ``OVERWRITE``, Velox will delete exising files if any and overwrite the partitions. The property will be set to ``ERROR`` if not specified, and raise error if writing to an existing partition.

Trino has similar semantics, https://trino.io/docs/current/connector/hive.html.

Yes, this property also applies to unpartitioned table. Cross-reading the presto java oss code gives me the impression that the unpartitioned table might be treated to be single partitioned.

gggrace14 commented on velox/connectors/hive/HiveConfig.h:
```diff
@@ -0,0 +1,55 @@
+      const Config* baseConfig);
+
+ private:
+  static InsertExistingPartitionsBehavior
```
Hi Masha, Do you mean I move the func to an anonymous namespace .cpp? I had moved the impl to .cpp file. This func is also called outside in HiveWriteProtocol.cpp.

gggrace14 replied:
Moved. Good to know

gggrace14 commented on velox/connectors/hive/HiveConfig.cpp:
```diff
@@ -0,0 +1,75 @@
+  if (strBehavior.has_value()) {
+    InsertExistingPartitionsBehavior behavior =
+        stringToInsertExistingPartitionsBehavior(strBehavior.value());
+    if (isImmutablePartitions(baseConfig)) {
```
This hive.immutable-partitions property is also shared w/ coordinator, and used extensively there. I suggest we keep it.

The logic I understand is, if it's configured to be immutable partitions, then if writing to an existing partition, for the three behaviors
APPEND -- not allowed
OVERWRITE -- allowed
ERROR -- raise error

if it's configed mutable partitions, then three behaviors
APPEND -- allowd
OVERWRITE -- allowd
ERROR -- raise error

This  immutable-partitions property and the behavior property are in many places across worker and the coordinator. Kind of their communication. I only understand them to this extent after revising several time.

gggrace14 commented:
Hi @mbasmanova , done w/ revising this PR and addressing the comments. Sorry, for missing the previous comments on the configs.rst. Might have forgotten to refresh the page?

gggrace14 commented on velox/connectors/hive/HiveDataSink.cpp:
```diff
@@ -17,41 +17,151 @@
+  columns.emplace_back(commitContextVector);
+
+  return std::make_shared<RowVector>(
+      pool, outputType_, nullptr, numOutputRows, columns);
```
The size of the three child vectors of the `TableWriter::getOutput()`, `numOutputRows` here, is the number of partitions plus 1. Only connector/HiveDataSink has this info, so I decide to assemble the three output vectors inside HiveDataSink instead of TableWriter.

I also need to pass Operator's `outputType_` to connector/DataSink/HiveDataSink, since `outputType_ `is used at the same place as `numOutputRows` as in line 164.

Also I move the logic to track the `numWrittenRows` from TableWriter to connector/HiveDataSink, since we actually need to track the per-partition `numWrittenRows`. numWrittenRows is also needed to populate the first vector.

gggrace14 reviewed (commented):
Hi @mbasmanova , this PR captures our discussion on how to refactor the APIs for the table writer. Ready for review :)

mbasmanova commented on velox/connectors/hive/HiveDataSink.cpp:
```diff
@@ -70,33 +123,86 @@ std::unique_ptr<velox::dwrf::Writer> HiveDataSink::createWriter() {
+  }
+
+  HiveWriterParameters::UpdateMode updateMode;
+  if (hiveTableWriteHandle->isInsertTable()) {
```
nit: this method is rather long; consider extracting this logic into a helper method

```
auto updateMode = getUpdateMode();
```

gggrace14 replied:
Helper func added

mbasmanova commented on velox/docs/configs.rst:
```diff
@@ -164,4 +164,27 @@ data size exceeds this limit, otherwise it spills the partition with most data.
+    * **Allowed values:** ``OVERWRITE``, ``ERROR``
+    * **Default value:** ``ERROR``
+
+Set the behavior on insert existing partitions in write protocol. ``OVERWRITE``
```
As discussed offline, this configuration property doesn't have any impact on the worker's behavior and is used only to populate the update-mode field in the writer's result. It would be nice to clarify this, otherwise, readers will expect some logic in the worker that doesn't exist.

gggrace14 replied:
Revised

mbasmanova commented on velox/exec/TableWriter.cpp:
```diff
@@ -53,8 +52,7 @@ TableWriter::TableWriter(
   mappedType_ = ROW(std::move(names), std::move(types));
-  writeProtocol_ = connector::WriteProtocol::getWriteProtocol(
-      tableWriteNode->commitStrategy());
+  commitStrategy_ = tableWriteNode->commitStrategy();
```
nit: consider making commitStrategy_ const and initialize it using initialization list

```
insertTableHandle_(
          tableWriteNode->insertTableHandle()->connectorInsertTableHandle()),
commitStrategy_(tableWriteNode->commitStrategy())
```

gggrace14 replied:
Revised this and others

mbasmanova commented on velox/exec/TableWriter.cpp:
```diff
@@ -99,11 +97,56 @@ RowVectorPtr TableWriter::getOutput() {
-  return writeProtocol_->commit(commitInfo, pool());
+  memory::MemoryPool* pool = connectorQueryCtx_->memoryPool();
+
+  if (outputType_->size() <= 1) {
```
Shouldn't this be == 1?

mbasmanova commented on velox/exec/TableWriter.cpp:
```diff
@@ -99,11 +97,56 @@ RowVectorPtr TableWriter::getOutput() {
+
+  std::vector<std::string> fragments = dataSink_->finish();
+
+  std::vector<VectorPtr> columns = {};
```
nit: consider moving this variable the end of the method and initialize it like so:

```
std::vector<VectorPtr> columns = {writtenRowsVector, fragmentsVector, commitContextVector};
```

mbasmanova commented on velox/connectors/hive/HiveDataSink.cpp:
```diff
@@ -70,33 +123,86 @@ std::unique_ptr<velox::dwrf::Writer> HiveDataSink::createWriter() {
+      break;
+    }
+    case CommitStrategy::kTaskCommit: {
+      targetFileName = fmt::format(
```
I'd like to get a better understanding of this naming scheme. In particular, it seems to me that taskId is globally unique, hence, taskId + driverId + 0 is unique. If that's the case, why is it necessary to add uuid when making writeFileName? Do we expect some garbage collection logic to look at files named .tmp.velox.xxx?

gggrace14 replied:
Hi @mbasmanova , the uuid part is to handle the case of task retry and speculative execution, so that the file names won't conflict during task retry. What I know about is that coordinator side (probably through warm storage service) will ignore files prefixed with dot, as been told many times.

gggrace14 commented on velox/connectors/Connector.h:
```diff
@@ -91,18 +88,34 @@ class ConnectorInsertTableHandle {
   // TODO maybe at some point we want to make it async
   virtual void appendData(VectorPtr input) = 0;
 
+  virtual std::vector<std::string> finish() const = 0;
```
Thanks for suggestion :)

gggrace14 commented on velox/connectors/hive/HiveDataSink.h:
```diff
@@ -105,39 +85,126 @@ class HiveInsertTableHandle : public ConnectorInsertTableHandle {
   // Parameters used by writers, and thus are tracked in the same order
   // as the writers_ vector
-  std::vector<std::shared_ptr<const HiveWriterParameters>> writerParameters_;
+  std::vector<std::shared_ptr<HiveWriterInfo>> writerInfo_;
```
I see. Will revisit it when I need to expose more metrics for the writer which will be soon. I also feel the stats, properties, etc. should be the encapsulated inside a "Writer" class.

gggrace14 commented on velox/connectors/hive/HiveDataSink.cpp:
```diff
@@ -17,41 +17,93 @@
-    writers_.emplace_back(createWriter());
+    std::shared_ptr<const HiveWriterParameters> hiveWriterParameters =
+        getWriterParameters(std::nullopt);
+    writers_.emplace_back(createWriter(hiveWriterParameters));
```
Make sense to me. I see.

gggrace14 commented on velox/connectors/hive/HiveDataSink.cpp:
```diff
@@ -17,41 +17,93 @@
+  partitionUpdates.reserve(writerInfo_.size());
+
+  for (const auto info : writerInfo_) {
+    if (info) {
```
Yeah, I think VectorHasher reserves 0 for all-null value/partition value. If the input doesn't have the all-null partition, then the writeInfo_ as well as the writers_ will be set to nullptr at index 0.

(I know it's not being used in this PR. Just do the check here in case I forget later.)

gggrace14 commented on velox/connectors/hive/HiveDataSink.cpp:
```diff
@@ -17,41 +17,93 @@
+          ("rowCount", info->numWrittenRows)
+         // TODO(gaoge): track and send the fields when inMemoryDataSizeInBytes, onDiskDataSizeInBytes
+         // and containsNumberedFileNames are needed at coordinator when file_renaming_enabled are turned on,
+         // https://fburl.com/18vim7k4.
```
Good catch.

gggrace14 commented on velox/connectors/hive/HiveDataSink.cpp:
```diff
@@ -70,33 +123,86 @@ std::unique_ptr<velox::dwrf::Writer> HiveDataSink::createWriter() {
+std::shared_ptr<const HiveWriterParameters> HiveDataSink::getWriterParameters(
+    const std::optional<std::string>& partition) const {
+  auto hiveTableWriteHandle =
+      std::dynamic_pointer_cast<const HiveInsertTableHandle>(
```
Good catch.

mbasmanova commented on velox/docs/configs.rst:
```diff
@@ -164,4 +164,29 @@ data size exceeds this limit, otherwise it spills the partition with most data.
+Hive Connector
+-----------------------------
+
+``max_partitions_per_writers``
```
As discussed, let's add this in the same PR that starts using it.

gggrace14 replied:
Forgot this place

mbasmanova commented on velox/docs/configs.rst:
```diff
@@ -164,4 +164,29 @@ data size exceeds this limit, otherwise it spills the partition with most data.
+    * **Allowed values:** ``OVERWRITE``, ``ERROR``
+    * **Default value:** ``ERROR``
+
+The behavior on insert existing partitions. This property only derives
```
Nice description.

mbasmanova commented on velox/exec/TableWriter.cpp:
```diff
@@ -99,11 +94,58 @@ RowVectorPtr TableWriter::getOutput() {
+  vector_size_t numOutputRows = fragments.size() + 1;
+
+  // Set rows column.
+  FlatVectorPtr<int64_t> writtenRowsVector =
```
nit: auto

mbasmanova commented on velox/exec/TableWriter.cpp:
```diff
@@ -99,11 +94,58 @@ RowVectorPtr TableWriter::getOutput() {
+  }
+
+  // Set fragments column.
+  FlatVectorPtr<StringView> fragmentsVector =
```
nit: auto

mbasmanova commented on velox/exec/TableWriter.cpp:
```diff
@@ -99,11 +94,58 @@ RowVectorPtr TableWriter::getOutput() {
+           ("pageSinkCommitStrategy", commitStrategyToString(commitStrategy_))
+           ("lastPage", true));
+  // clang-format on
+  VectorPtr commitContextVector = BaseVector::createConstant(
```
nit: auto

mbasmanova commented on velox/exec/TableWriter.cpp:
```diff
@@ -99,11 +94,58 @@ RowVectorPtr TableWriter::getOutput() {
+      variant::binary(commitContextJson), numOutputRows, pool);
+
+  std::vector<VectorPtr> columns = {
+      writtenRowsVector, fragmentsVector, commitContextVector};
```
commitContextVector is a constant varbinary vector. We need to make sure it is not flattened during serialization (PrestoSerializer) and passing from C++ process to Java process (running TableFinish operator). Otherwise, the JSON string will get copied many times. Let's make sure to follow-up.

gggrace14 replied:
I need to follow up on this

mbasmanova reviewed (approved):
@gggrace14 Thank you. Looks good to me. Please, update the PR description to (1) remove max_partitions_per_writers; (2) mention that fileSize is currently not populated (also, please add a TODO to the code); (3) mention a need to follow up re: serialization of the table write result to make sure constant vector of commit context doesn't get flattened.
